pyquil/api/_qpu.py

```diff
@@ -197,15 +206,28 @@ def run(self, run_priority: Optional[int] = None) -> "QPU":
 
         return self
 
+    @contextmanager
+    def _qpu_client(self) -> Iterator[QPUClient]:
```

Comment:
Technically, there's a remote possibility that the engagement expires between retrieving this client and actually using it. Since the client itself is stateless, it has no mechanism to detect this.

I don't think it's a big enough problem to worry about, but there could be some instances where users get errors due to them executing a program _just before_ any current engagement they have expires.